### ISSUE TITLE
issue with simple line drawing 

### ISSUE THREAD

**Issue by stephanschulz** [openframeworks/openFrameworks#6001]
I’m on macOS 10.12.6 with Xcode Version 9.2 beta (9C32c) with MBP Retina display
running this simple code

```
    ofBackground(0);
    int _h = mouseX;
    ofSetColor(255);
    for(int i=0; i<1000; i++){
        
        ofDrawLine(i, _h,i, _h+20);
    }
```

produces this image where lines should have been drawn one beside the next but for some reason some are on top of each other. I have seen this in previous versions too.

<img width="1015" alt="screen shot 2018-05-04 at 9 23 15 am" src="https://user-images.githubusercontent.com/1054816/39635016-467e90f0-4f8a-11e8-8b44-8859fe646ec2.png">

**Comment by m-uv** [comment]
I had a similar issue a few months ago on a MacBook pro (15inch with retina display from mid 2015), on macOS Sierra 10.12.6. I tried the code on Xcode Version 9.2 (9C40b), Xcode Version 8.0 (8A218a) and Qt Creator as well as different versions of openFrameworks (of_v0.9.8_osx, of_v20170714_osx, 0.10.0) I just run it again trying the new of0.10.4 on Xcode Version 9.2 (9C40b) with the same results. 

This was the code to create a gradient by drawing vertical lines:

```
 
    ofBackground(0);
    for (int i = 0; i < ofGetWidth(); i++)
    {
        int grey = ofWrap(i, 0, 255);
        ofSetColor(grey);
        ofDrawLine(i, 0, i, 255);
    }

```

And this was the output
<img width="256" alt="screen shot 2018-05-04 at 12 55 26 pm" src="https://user-images.githubusercontent.com/22039749/39643809-9025048a-4f9b-11e8-9973-fae70d1e3a99.png">

I was able to “fake” the desired output by setting the lineWidth to 2. 

```  

    ofBackground(0);
    for (int i = 0; i < ofGetWidth(); i++)
    {
        int grey = ofWrap(i, 0, 255);
        ofSetColor(grey);
	ofSetLineWidth(2);
        ofDrawLine(i, 0, i, 255);
    }

```

This is the result:
<img width="259" alt="screen shot 2018-05-04 at 12 56 08 pm" src="https://user-images.githubusercontent.com/22039749/39643847-a839ae18-4f9b-11e8-86a0-c25c84c2a47f.png">

However, when drawing horizontal lines the issue did not occur, although the gradient did not go all the way to white (which it did when setting the line width to 2 again):

```  

    ofBackground(0);
    for (int i = 0; i < ofGetWidth(); i++)
    {
        int grey = ofWrap(i, 0, 255);
        ofSetColor(grey);
        ofDrawLine(0, i, 255, i);
    }

```
<img width="263" alt="screen shot 2018-05-04 at 12 57 02 pm" src="https://user-images.githubusercontent.com/22039749/39643892-c918af44-4f9b-11e8-8367-e326a47af4ca.png">

**Comment by bakercp** [comment]
To add another data point:
- macOS 10.13.4
- MacBook Pro (15-inch, 2017) (retina)
- Xcode Version 9.2 (9C40b)

```cpp
 ofApp::draw() {
    ofBackground(0);
    for (int i = 0; i < ofGetWidth(); i++)
    {
        int grey = ofWrap(i, 0, 255);
        ofSetColor(grey);
        ofDrawLine(i, 0, i, 255);
    }
}
```

<img width="1028" alt="ofapp_cpp" src="https://user-images.githubusercontent.com/300484/39645620-9954ddc0-4f9e-11e8-9c86-7f4485e6f795.png">

**Comment by bakercp** [comment]
To add another data point:

oF Master (RC4)
macOS 10.13.4
MacBook Pro (15-inch, 2017) (retina)
QTCreator 4.5.1

<img width="512" alt="blank_skitch_document" src="https://user-images.githubusercontent.com/300484/39645807-12099a12-4f9f-11e8-8581-26ad8df0290f.png">

I also tried to open the same app in low res mode with the same results.
<img width="400" alt="test_debug_app_info" src="https://user-images.githubusercontent.com/300484/39645872-44e8568a-4f9f-11e8-8ed2-f8e73a429fea.png">

![blank_skitch_document](https://user-images.githubusercontent.com/300484/39645831-24e941a0-4f9f-11e8-972c-e4704ab827d0.png)

**Comment by bakercp** [comment]
Also, it doesn't seem to matter what GL version is used `main.cpp`:

```cpp
#include "ofApp.h"

int main()
{
    ofGLWindowSettings settings;
    settings.setSize(1024, 768);
//    settings.setGLVersion(2, 1);
//    settings.setGLVersion(3, 2);
//    settings.setGLVersion(4, 1);
    settings.windowMode = OF_WINDOW;
    auto window = ofCreateWindow(settings);
    auto app = std::make_shared<ofApp>();
    ofRunApp(window, app);
    return ofRunMainLoop();
}
```

**Comment by arturoc** [comment]
Have you tried disabling antialiasing? in main set window settings numSamples to 0 or in setup call ofDisableAntialiasing()

**Comment by jvcleave** [comment]
This is with 0.9.8 but there are differences in Release/Debug so may offer a clue

Release:
![image](https://user-images.githubusercontent.com/150037/39661257-890a1b32-501c-11e8-85cd-956c9f9e6aa5.png)

 Debug:
![image](https://user-images.githubusercontent.com/150037/39661263-a277e7b6-501c-11e8-82ad-e6329d8b6d88.png)

**Comment by bakercp** [comment]
@arturoc `ofDisableSmoothing()` has no effect.

Playing with `numSamples` also doesn't seem to have an effect.

```cpp
int main()
{
    ofGLFWWindowSettings settings;
    settings.setSize(1024, 768);
    //    settings.setGLVersion(4, 1);
    settings.numSamples = 4;
    settings.windowMode = OF_WINDOW;
    auto window = ofCreateWindow(settings);
    auto app = std::make_shared<ofApp>();
    ofRunApp(window, app);
    return ofRunMainLoop();
}
```

**Comment by bakercp** [comment]
Could this be a GL driver bug on macOS?

**Comment by arturoc** [comment]
no, ofDisableSmoothing is a different thing and is disabled by default, i was saying ofDisableAntialiasing() but numSamples=0 should have the same effect

**Comment by tgfrerer** [comment]
@bakercp looks like a GL driver issue to me. It could be that with Mac OS Sierra and onwards OpenGL calls are routed/emulated through Metal, and that line drawing + overdraw trip up their driver. A way to check this hypothesis would be to test the code on a Mac with older OS...

For comparison, here's how linux draws the above:

![screenshot from 2018-05-10 12-00-15](https://user-images.githubusercontent.com/423509/39866754-d5d199b4-5449-11e8-90bf-a4f777fa542d.png)

**Comment by bakercp** [comment]
I suspect this has something to do with AMD graphics cards.

**Comment by dimitre** [comment]
+1 for this amd suspect
no artifacts on M1 here

**Comment by dimitre** [comment]
I suppose this issue won't be addressed because it seems to be OS related.
should we close it?
